src/core/browser/qgsfilebaseddataitemprovider.cpp

```diff
@@ -0,0 +1,353 @@
+{
+  if ( QgsProviderUtils::sublayerDetailsAreIncomplete( mSublayers, QgsProviderUtils::SublayerCompletenessFlag::IgnoreUnknownFeatureCount ) )
+  {
+    mSublayers = QgsProviderRegistry::instance()->querySublayers( path(), Qgis::SublayerQueryFlag::ResolveGeometryType );
+  }
+
+  QVector<QgsDataItem *> children;
+  children.reserve( mSublayers.size() );
+  for ( const QgsProviderSublayerDetails &sublayer : std::as_const( mSublayers ) )
+  {
+    QgsProviderSublayerItem *item = new QgsProviderSublayerItem( this, sublayer.name(), sublayer, true );
+    children.append( item );
+  }
+
+  return children;
+}
+
+bool QgsFileDataCollectionItem::hasDragEnabled() const
+{
+  return true;
+}
+
+QgsMimeDataUtils::UriList QgsFileDataCollectionItem::mimeUris() const
+{
+  QgsMimeDataUtils::Uri collectionUri;
+  collectionUri.uri = path();
+  collectionUri.layerType = QStringLiteral( "collection" );
+  return { collectionUri };
+}
+
+QgsAbstractDatabaseProviderConnection *QgsFileDataCollectionItem::databaseConnection() const
+{
+  // sqlite gets special handling because of the spatialite provider which supports the api required database connections
+  const QFileInfo fi( mPath );
+  if ( fi.suffix().toLower() != QLatin1String( "sqlite" )  && fi.suffix().toLower() != QLatin1String( "db" ) )
+  {
+    return nullptr;
+  }
+
+  QgsAbstractDatabaseProviderConnection *conn = nullptr;
+
+  // test that file is valid with OGR
+  if ( OGRGetDriverCount() == 0 )
+  {
+    OGRRegisterAll();
+  }
+  // do not print errors, but write to debug
+  CPLPushErrorHandler( CPLQuietErrorHandler );
+  CPLErrorReset();
+  gdal::dataset_unique_ptr hDS( GDALOpenEx( path().toUtf8().constData(), GDAL_OF_VECTOR, nullptr, nullptr, nullptr ) );
+  CPLPopErrorHandler();
+
+  if ( ! hDS )
+  {
+    QgsDebugMsgLevel( QStringLiteral( "GDALOpen error # %1 : %2 on %3" ).arg( CPLGetLastErrorNo() ).arg( CPLGetLastErrorMsg() ).arg( path() ), 2 );
+    return nullptr;
+  }
+
+  GDALDriverH hDriver = GDALGetDatasetDriver( hDS.get() );
+  QString driverName = GDALGetDriverShortName( hDriver );
+
+  if ( driverName == QLatin1String( "SQLite" ) )
+  {
+    QgsProviderMetadata *md { QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "spatialite" ) ) };
+    if ( md )
+    {
+      QgsDataSourceUri uri;
+      uri.setDatabase( path( ) );
+      conn = static_cast<QgsAbstractDatabaseProviderConnection *>( md->createConnection( uri.uri(), {} ) );
+    }
+  }
+  return conn;
+}
+
+//
+// QgsFileBasedDataItemProvider
+//
+
+QString QgsFileBasedDataItemProvider::name()
+{
+  return QStringLiteral( "files" );
+}
+
+int QgsFileBasedDataItemProvider::capabilities() const
+{
+  return QgsDataProvider::File | QgsDataProvider::Dir;
+}
+
+QgsDataItem *QgsFileBasedDataItemProvider::createDataItem( const QString &pathIn, QgsDataItem *parentItem )
+{
+  QString path( pathIn );
+  if ( path.isEmpty() )
+    return nullptr;
+
+  const QFileInfo info( path );
+  QString suffix = info.suffix().toLower();
```

Comment:
```suggestion
  const QString suffix = info.suffix().toLower();
```
?

Reply:
it's modified in some code paths, so isn't const